sdk/keyvault/azure-keyvault-keys/azure/keyvault/keys/crypto/_models.py

```diff
@@ -2,11 +2,363 @@
+    if isinstance(padding, PSS):
+        mapped_algorithm = PSS_MAP.get(mapped_algorithm)
+
+        # Public mgf property was only added in https://github.com/pyca/cryptography/pull/9582
+        # _mgf property has been available in every version of the PSS class, so we use it as a backup
+        try:
+            mgf = padding.mgf  # type: ignore[attr-defined]
+        except AttributeError:
+            mgf = padding._mgf  # pylint:disable=protected-access
+        if not isinstance(mgf, MGF1):
+            raise ValueError(f"Unsupported MGF: {mgf}")
+
+    # The only other padding accepted is PKCS1v15
+    elif not isinstance(padding, PKCS1v15):
+        raise ValueError(f"Unsupported padding: {padding.name}")
+
+    return cast(SignatureAlgorithm, mapped_algorithm)
+
+
+class KeyVaultRSAPublicKey(RSAPublicKey):
+    """An `RSAPublicKey` implementation based on a key managed by Key Vault."""
+
+    def __init__(self, client: "CryptographyClient", key_material: JsonWebKey) -> None:
+        """Creates a `KeyVaultRSAPublicKey` from a `CryptographyClient` and key.
+
+        :param client: The client that will be used to communicate with Key Vault.
+        :type client: :class:`~azure.keyvault.keys.crypto.CryptographyClient`
+        :param key_material: They Key Vault key's material, as a `JsonWebKey`.
+        :type key_material: :class:`~azure.keyvault.keys.JsonWebKey`
+        """
+        self._client: "CryptographyClient" = client
+        self._key: JsonWebKey = key_material
+
+    def encrypt(self, plaintext: bytes, padding: AsymmetricPadding) -> bytes:
+        """Encrypts the given plaintext.
+
+        :param bytes plaintext: Plaintext to encrypt.
+        :param padding: The padding to use. Supported paddings are `OAEP` and `PKCS1v15`. For `OAEP` padding, supported
+            hash algorithms are `SHA1` and `SHA256`. The only supported mask generation function is `MGF1`. See
+            https://learn.microsoft.com/azure/key-vault/keys/about-keys-details for details.
+        :type padding: AsymmetricPadding
+
+        :returns: The encrypted ciphertext, as bytes.
+        :rtype: bytes
+        """
+        mapped_algorithm = get_encryption_algorithm(padding)
+        result = self._client.encrypt(mapped_algorithm, plaintext)
+        return result.ciphertext
+
+    @property
+    def key_size(self) -> int:
+        """The bit length of the public modulus.
+
+        :returns: The key's size.
+        :rtype: int
+        """
+        return len(self._key.n) * 8  # type: ignore[attr-defined]
+
+    def public_numbers(self) -> RSAPublicNumbers:
+        """Returns an `RSAPublicNumbers` representing the key's public numbers.
+
+        :returns: The public numbers of the key.
+        :rtype: RSAPublicNumbers
+        """
+        e = int.from_bytes(self._key.e, "big")  # type: ignore[attr-defined]
+        n = int.from_bytes(self._key.n, "big")  # type: ignore[attr-defined]
+        return RSAPublicNumbers(e, n)
```

Comment:
I see the `RSAPublicNumbers` object has it's own `public_key()` method returning `RSAPublicKey` (same for `RSAPrivateNumbers`).
It would be good to test what this does... e.g. if I have code like:
```python
my_private_key = client.create_rsa_private_key()
my_public_key = my_private_key.private_numbers().public_key()

my_public_key.verify(b"foo", b"bar") == client.create_rsa_public_key.verify(b"foo", b"bar")
```

Reply:
This actually solves a couple of problems! I wasn't sure how to go about implementing some methods, like `public_bytes` or `recover_data_from_signature` -- but fetching a key from `public_numbers` or `private_numbers` actually lets us just use those implementations. That makes sense to do anyway, since they're not using any functionality from Key Vault. Thanks for the suggestion!

Reply:
No, unfortunately. Since the base class is abstract, we can't just use `super()` directly -- but going through `PublicNumbers`/`PrivateNumbers` to _then_ create an equivalent key works because actual OpenSSL implementations are used to return functional keys: https://github.com/pyca/cryptography/blob/4553a409e30d280e8f773921db8fdd71931fe230/src/cryptography/hazmat/primitives/asymmetric/rsa.py#L427-L429